File: typescript/src/resources/discord-bot.ts
/** Discord Agent Proxy client.
 *
 * The Discord Agent Proxy is an independently deployed service.  Users
 * deploy it from the AceData Cloud console at
 * https://platform.acedata.cloud/console/applications and receive a
 * deployment-specific base URL and access token.
 *
 * Use {@link DiscordBotClient} to interact with the REST API exposed by
 * that deployment.
 */

export interface DiscordBotClientOptions {
  /** Deployment URL, e.g. `https://discord-bot-xxxxxxxxxxxx.app.acedata.cloud` */
  baseURL: string;
  /** Access token shown in the console for this deployment */
  token: string;
  /** Request timeout in milliseconds (default 300_000) */
  timeout?: number;
}

export class DiscordBotClient {
  private base: string;
  private token: string;
  private timeout: number;

  constructor(opts: DiscordBotClientOptions) {
    this.base = opts.baseURL.replace(/\/$/, '');
    this.token = opts.token;
    this.timeout = opts.timeout ?? 300_000;
  }

  private async request<T = unknown>(
    method: string,
    path: string,
    opts?: { json?: unknown; params?: Record<string, string | number> },
  ): Promise<T> {
    let url = `${this.base}${path}`;
    if (opts?.params) {
      const qs = new URLSearchParams(
        Object.entries(opts.params).map(([k, v]): [string, string] => [k, String(v)]),
      );
      url = `${url}?${qs}`;
    }
    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), this.timeout);
    try {
      const resp = await fetch(url, {
        method,
        headers: {
          accept: 'application/json',
          'content-type': 'application/json',
          authorization: this.token,
        },
        body: opts?.json != null ? JSON.stringify(opts.json) : undefined,
        signal: controller.signal,
      });
      const body = (await resp.json()) as Record<string, unknown>;
      if (!resp.ok) {
        const msg = typeof body.error === 'string' ? body.error : resp.statusText;
        const err = new Error(msg) as Error & { statusCode: number };
        err.statusCode = resp.status;
        throw err;
      }
      return ('data' in body ? body.data : body) as T;
    } finally {
      clearTimeout(id);
    }
  }

  // ── No-auth endpoint ───────────────────────────────────────────────

  /** Check service health (no auth required). */
  async health(): Promise<{ status: string; gateway_ready: boolean }> {
    const resp = await fetch(`${this.base}/health`);
    return resp.json() as Promise<{ status: string; gateway_ready: boolean }>;
  }

  // ── Account ────────────────────────────────────────────────────────

  /** Return information about the proxied Discord account. */
  async whoami(): Promise<Record<string, unknown>> {
    return this.request('GET', '/api/whoami');
  }

  // ── Guilds ─────────────────────────────────────────────────────────

  /** List all guilds (servers) the account has joined. */
  async listGuilds(): Promise<unknown[]> {
    return this.request('GET', '/api/guilds');
  }

  /** List channels in a guild. */
  async listChannels(guildId: string): Promise<unknown[]> {
    return this.request('GET', `/api/guilds/${guildId}/channels`);
  }

  /** Create a text channel in a guild. */
  async createChannel(
    guildId: string,
    opts: { name: string },
  ): Promise<Record<string, unknown>> {
    return this.request('POST', `/api/guilds/${guildId}/channels`, { json: opts });
  }

  /** List members of a guild (default limit 100). */
  async listMembers(
    guildId: string,
    opts?: { limit?: number },
  ): Promise<unknown[]> {
    return this.request('GET', `/api/guilds/${guildId}/members`, {
      params: opts?.limit != null ? { limit: opts.limit } : undefined,
    });
  }

  // ── Messages ───────────────────────────────────────────────────────

  /** Send a message to a channel. */
  async sendMessage(opts: {
    channel_id: string;
    content: string;
    reply_to?: string;
  }): Promise<Record<string, unknown>> {
    return this.request('POST', '/api/messages', { json: opts });
  }

  /** Read recent messages from a channel (default 50, max 100). */
  async listMessages(
    channelId: string,
    opts?: { limit?: number },
  ): Promise<unknown[]> {
    return this.request('GET', `/api/channels/${channelId}/messages`, {
      params: opts?.limit != null ? { limit: opts.limit } : undefined,
    });
  }

  /** Search messages in a channel. */
  async searchMessages(
    channelId: string,
    opts: { q: string; limit?: number },
  ): Promise<unknown[]> {
    return this.request('GET', `/api/channels/${channelId}/messages/search`, {
      params: {
        q: opts.q,
        ...(opts.limit != null ? { limit: opts.limit } : {}),
      },
    });
  }

  /** Edit a message you own. */
  async editMessage(
    channelId: string,
    messageId: string,
    opts: { content: string },
  ): Promise<Record<string, unknown>> {
    return this.request('PATCH', `/api/channels/${channelId}/messages/${messageId}`, {
      json: opts,
    });
  }

  /** Delete a message. */
  async deleteMessage(
    channelId: string,
    messageId: string,
  ): Promise<Record<string, unknown>> {
    return this.request('DELETE', `/api/channels/${channelId}/messages/${messageId}`);
  }

  /** Add an emoji reaction to a message. */
  async addReaction(
    channelId: string,
    messageId: string,
    opts: { emoji: string },
  ): Promise<Record<string, unknown>> {
    return this.request(
      'POST',
      `/api/channels/${channelId}/messages/${messageId}/reactions`,
      { json: opts },
    );
  }

  /** Pin a message in a channel. */
  async pinMessage(
    channelId: string,
    messageId: string,
  ): Promise<Record<string, unknown>> {
    return this.request('POST', `/api/channels/${channelId}/messages/${messageId}/pin`);
  }

  // ── Direct messages ────────────────────────────────────────────────

  /** Open a DM channel with a user; returns the channel info. */
  async createDm(opts: { recipient_id: string }): Promise<Record<string, unknown>> {
    return this.request('POST', '/api/dms', { json: opts });
  }

  /** Send a direct message to a user. */
  async sendDm(opts: {
    recipient_id: string;
    content: string;
  }): Promise<Record<string, unknown>> {
    return this.request('POST', '/api/dms/send', { json: opts });
  }
}

File: python/src/acedatacloud/resources/discord_bot.py
"""Discord Agent Proxy client.

The Discord Agent Proxy is an independently deployed service.  Users
deploy it from the AceData Cloud console at
https://platform.acedata.cloud/console/applications and receive a
deployment-specific base URL and access token.

This module provides :class:`DiscordBotClient` (synchronous) and
:class:`AsyncDiscordBotClient` (async) for interacting with the REST
API exposed by that deployment.
"""

from __future__ import annotations

from typing import Any

import httpx


class _DiscordBotBase:
    def __init__(self, base_url: str, token: str, *, timeout: float = 300.0) -> None:
        self._base = base_url.rstrip("/")
        self._token = token
        self._timeout = timeout

    def _make_headers(self) -> dict[str, str]:
        return {
            "accept": "application/json",
            "content-type": "application/json",
            "authorization": self._token,
        }

    @staticmethod
    def _unwrap(body: Any) -> Any:
        if isinstance(body, dict) and "data" in body:
            return body["data"]
        return body


class DiscordBotClient(_DiscordBotBase):
    """Synchronous client for a deployed Discord Agent Proxy.

    Args:
        base_url: The deployment URL shown in the AceData Cloud console,
            e.g. ``https://discord-bot-xxxxxxxxxxxx.app.acedata.cloud``.
        token: The access token for this deployment.
        timeout: HTTP timeout in seconds (default 300).
    """

    def __init__(self, base_url: str, token: str, *, timeout: float = 300.0) -> None:
        super().__init__(base_url, token, timeout=timeout)
        self._client = httpx.Client(timeout=timeout)

    def _request(self, method: str, path: str, **kwargs: Any) -> Any:
        url = f"{self._base}{path}"
        resp = self._client.request(method, url, headers=self._make_headers(), **kwargs)
        resp.raise_for_status()
        return self._unwrap(resp.json())

    # ── No-auth endpoint ──────────────────────────────────────────────

    def health(self) -> dict[str, Any]:
        """Check service health.  Does not require authentication."""
        url = f"{self._base}/health"
        resp = self._client.get(url)
        resp.raise_for_status()
        return resp.json()  # type: ignore[return-value]

    # ── Account ───────────────────────────────────────────────────────

    def whoami(self) -> dict[str, Any]:
        """Return information about the proxied Discord account."""
        return self._request("GET", "/api/whoami")  # type: ignore[return-value]

    # ── Guilds ────────────────────────────────────────────────────────

    def list_guilds(self) -> list[Any]:
        """List all guilds (servers) the account has joined."""
        return self._request("GET", "/api/guilds")  # type: ignore[return-value]

    def list_channels(self, guild_id: str) -> list[Any]:
        """List channels in a guild."""
        return self._request("GET", f"/api/guilds/{guild_id}/channels")  # type: ignore[return-value]

    def create_channel(self, guild_id: str, *, name: str) -> dict[str, Any]:
        """Create a text channel in a guild."""
        return self._request(  # type: ignore[return-value]
            "POST", f"/api/guilds/{guild_id}/channels", json={"name": name}
        )

    def list_members(self, guild_id: str, *, limit: int = 100) -> list[Any]:
        """List members of a guild (default 100)."""
        return self._request(  # type: ignore[return-value]
            "GET", f"/api/guilds/{guild_id}/members", params={"limit": limit}
        )

    # ── Messages ──────────────────────────────────────────────────────

    def send_message(
        self, *, channel_id: str, content: str, reply_to: str | None = None
    ) -> dict[str, Any]:
        """Send a message to a channel.

        Args:
            channel_id: Target channel ID.
            content: Message text.
            reply_to: Optional message ID to reply to.
        """
        body: dict[str, Any] = {"channel_id": channel_id, "content": content}
        if reply_to is not None:
            body["reply_to"] = reply_to
        return self._request("POST", "/api/messages", json=body)  # type: ignore[return-value]

    def list_messages(self, channel_id: str, *, limit: int = 50) -> list[Any]:
        """Read recent messages from a channel (default 50, max 100)."""
        return self._request(  # type: ignore[return-value]
            "GET", f"/api/channels/{channel_id}/messages", params={"limit": limit}
        )

    def search_messages(self, channel_id: str, *, q: str, limit: int = 25) -> list[Any]:
        """Search messages in a channel."""
        return self._request(  # type: ignore[return-value]
            "GET",
            f"/api/channels/{channel_id}/messages/search",
            params={"q": q, "limit": limit},
        )

    def edit_message(
        self, channel_id: str, message_id: str, *, content: str
    ) -> dict[str, Any]:
        """Edit a message you own."""
        return self._request(  # type: ignore[return-value]
            "PATCH",
            f"/api/channels/{channel_id}/messages/{message_id}",
            json={"content": content},
        )

    def delete_message(self, channel_id: str, message_id: str) -> dict[str, Any]:
        """Delete a message."""
        return self._request(  # type: ignore[return-value]
            "DELETE", f"/api/channels/{channel_id}/messages/{message_id}"
        )

    def add_reaction(
        self, channel_id: str, message_id: str, *, emoji: str
    ) -> dict[str, Any]:
        """Add an emoji reaction to a message."""
        return self._request(  # type: ignore[return-value]
            "POST",
            f"/api/channels/{channel_id}/messages/{message_id}/reactions",
            json={"emoji": emoji},
        )

    def pin_message(self, channel_id: str, message_id: str) -> dict[str, Any]:
        """Pin a message in a channel."""
        return self._request(  # type: ignore[return-value]
            "POST", f"/api/channels/{channel_id}/messages/{message_id}/pin"
        )

    # ── Direct messages ───────────────────────────────────────────────

    def create_dm(self, *, recipient_id: str) -> dict[str, Any]:
        """Open a DM channel with a user; returns the channel info."""
        return self._request(  # type: ignore[return-value]
            "POST", "/api/dms", json={"recipient_id": recipient_id}
        )

    def send_dm(self, *, recipient_id: str, content: str) -> dict[str, Any]:
        """Send a direct message to a user."""
        return self._request(  # type: ignore[return-value]
            "POST", "/api/dms/send", json={"recipient_id": recipient_id, "content": content}
        )

    # ── Lifecycle ─────────────────────────────────────────────────────

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> DiscordBotClient:
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()


class AsyncDiscordBotClient(_DiscordBotBase):
    """Asynchronous client for a deployed Discord Agent Proxy.

    Args:
        base_url: The deployment URL shown in the AceData Cloud console,
            e.g. ``https://discord-bot-xxxxxxxxxxxx.app.acedata.cloud``.
        token: The access token for this deployment.
        timeout: HTTP timeout in seconds (default 300).
    """

    def __init__(self, base_url: str, token: str, *, timeout: float = 300.0) -> None:
        super().__init__(base_url, token, timeout=timeout)
        self._client = httpx.AsyncClient(timeout=timeout)

    async def _request(self, method: str, path: str, **kwargs: Any) -> Any:
        url = f"{self._base}{path}"
        resp = await self._client.request(method, url, headers=self._make_headers(), **kwargs)
        resp.raise_for_status()
        return self._unwrap(resp.json())

    async def health(self) -> dict[str, Any]:
        """Check service health.  Does not require authentication."""
        url = f"{self._base}/health"
        resp = await self._client.get(url)
        resp.raise_for_status()
        return resp.json()  # type: ignore[return-value]

    async def whoami(self) -> dict[str, Any]:
        """Return information about the proxied Discord account."""
        return await self._request("GET", "/api/whoami")  # type: ignore[return-value]

    async def list_guilds(self) -> list[Any]:
        """List all guilds (servers) the account has joined."""
        return await self._request("GET", "/api/guilds")  # type: ignore[return-value]

    async def list_channels(self, guild_id: str) -> list[Any]:
        """List channels in a guild."""
        return await self._request(  # type: ignore[return-value]
            "GET", f"/api/guilds/{guild_id}/channels"
        )

    async def create_channel(self, guild_id: str, *, name: str) -> dict[str, Any]:
        """Create a text channel in a guild."""
        return await self._request(  # type: ignore[return-value]
            "POST", f"/api/guilds/{guild_id}/channels", json={"name": name}
        )

    async def list_members(self, guild_id: str, *, limit: int = 100) -> list[Any]:
        """List members of a guild (default 100)."""
        return await self._request(  # type: ignore[return-value]
            "GET", f"/api/guilds/{guild_id}/members", params={"limit": limit}
        )

    async def send_message(
        self, *, channel_id: str, content: str, reply_to: str | None = None
    ) -> dict[str, Any]:
        """Send a message to a channel."""
        body: dict[str, Any] = {"channel_id": channel_id, "content": content}
        if reply_to is not None:
            body["reply_to"] = reply_to
        return await self._request("POST", "/api/messages", json=body)  # type: ignore[return-value]

    async def list_messages(self, channel_id: str, *, limit: int = 50) -> list[Any]:
        """Read recent messages from a channel (default 50, max 100)."""
        return await self._request(  # type: ignore[return-value]
            "GET", f"/api/channels/{channel_id}/messages", params={"limit": limit}
        )

    async def search_messages(
        self, channel_id: str, *, q: str, limit: int = 25
    ) -> list[Any]:
        """Search messages in a channel."""
        return await self._request(  # type: ignore[return-value]
            "GET",
            f"/api/channels/{channel_id}/messages/search",
            params={"q": q, "limit": limit},
        )

    async def edit_message(
        self, channel_id: str, message_id: str, *, content: str
    ) -> dict[str, Any]:
        """Edit a message you own."""
        return await self._request(  # type: ignore[return-value]
            "PATCH",
            f"/api/channels/{channel_id}/messages/{message_id}",
            json={"content": content},
        )

    async def delete_message(self, channel_id: str, message_id: str) -> dict[str, Any]:
        """Delete a message."""
        return await self._request(  # type: ignore[return-value]
            "DELETE", f"/api/channels/{channel_id}/messages/{message_id}"
        )

    async def add_reaction(
        self, channel_id: str, message_id: str, *, emoji: str
    ) -> dict[str, Any]:
        """Add an emoji reaction to a message."""
        return await self._request(  # type: ignore[return-value]
            "POST",
            f"/api/channels/{channel_id}/messages/{message_id}/reactions",
            json={"emoji": emoji},
        )

    async def pin_message(self, channel_id: str, message_id: str) -> dict[str, Any]:
        """Pin a message in a channel."""
        return await self._request(  # type: ignore[return-value]
            "POST", f"/api/channels/{channel_id}/messages/{message_id}/pin"
        )

    async def create_dm(self, *, recipient_id: str) -> dict[str, Any]:
        """Open a DM channel with a user; returns the channel info."""
        return await self._request(  # type: ignore[return-value]
            "POST", "/api/dms", json={"recipient_id": recipient_id}
        )

    async def send_dm(self, *, recipient_id: str, content: str) -> dict[str, Any]:
        """Send a direct message to a user."""
        return await self._request(  # type: ignore[return-value]
            "POST", "/api/dms/send", json={"recipient_id": recipient_id, "content": content}
        )

    async def close(self) -> None:
        await self._client.aclose()

    async def __aenter__(self) -> AsyncDiscordBotClient:
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self.close()
